src/envoy/http/authn/authn_utils.cc

```diff
@@ -100,6 +104,42 @@ bool AuthnUtils::ProcessJwtPayload(const std::string& payload_str,
   return true;
 }
 
+bool AuthnUtils::ExtractOriginalPayload(const std::string& token,
+                                        std::string* original_payload) {
+  Envoy::Json::ObjectSharedPtr json_obj;
+  try {
+    json_obj = Json::Factory::loadFromString(token);
+  } catch (...) {
+    return false;
+  }
+
+  if (json_obj->hasObject(kAPTokenOriginalPayload) == false) {
+    return false;
+  }
+
+  Envoy::Json::ObjectSharedPtr original_payload_obj;
+  try {
+    auto original_payload_obj = json_obj->getObject(kAPTokenOriginalPayload);
+    std::string iss1 = json_obj->getString(kJwtIssuerKey, "");
+    std::string iss2 = original_payload_obj->getString(kJwtIssuerKey, "");
+    // Token exchange makes the issuers of the APToken and the original JWT
```

Comment:
This logic sounds like a hack. It basically says "if original_payload exists, and original issuer is not the same as new issuer, it is AP token". This is not how AP token is defined.

Reply:
Based on the discussion today, the code has been revised.

Reply:
Do we still need to compare issuer?

Reply:
I recommend comparing issuers because if an token exchange occurs, the issuer of the exchanged token should be different from the issuer of the original token. @quanjielin @liminw, what is your thought on comparing issuers?

Reply:
I don't think it is a hard requirement that the issuers are different. We can probably drop the check here.